mirgecom/boundary.py

```diff
@@ -602,3 +689,628 @@ def temperature_bc(self, state_minus, **kwargs):
+
+        cv_infinity = make_conserved(
+            state_minus.dim, mass=free_stream_density,
+            energy=free_stream_total_energy,
+            momentum=free_stream_density*free_stream_velocity,
+            species_mass=free_stream_spec_mass
+        )
+
+        return make_fluid_state(cv=cv_infinity, gas_model=gas_model,
+                                temperature_seed=free_stream_temperature)
+
+    def temperature_bc(self, state_minus, **kwargs):
+        """Get temperature value to weakly prescribe flow temperature at boundary."""
+        return 0*state_minus.temperature + self._temperature
+
+
+class OutflowBoundary(PrescribedFluidBoundary):
+    r"""Outflow boundary treatment.
+
+    This class implements an outflow boundary as described by
+    [Mengaldo_2014]_.  The boundary condition is implemented
+    as:
+
+    .. math:
+
+        \rho^+ &= \rho^-
+        \rho\mathbf{Y}^+ &= \rho\mathbf{Y}^-
+        \rho\mathbf{V}^+ &= \rho^\mathbf{V}^-
+
+    Total energy for the flow is computed as follows:
+
+
+    When the flow is super-sonic, i.e. when:
+
+    .. math:
+
+       \rho\mathbf{V} \cdot \hat\mathbf{n} \ge c,
+
+    then the internal solution is used outright:
+
+    .. math:
+
+        \rho{E}^+ &= \rho{E}^-
+
+    otherwise the flow is sub-sonic, and the prescribed boundary pressure,
+    $P^+$, is used to compute the energy:
+
+    .. math:
+
+        \rho{E}^+ &= \frac{\left(2~P^+ - P^-\right)}{\left(\gamma-1\right)}
+        + \frac{1}{2\rho^+}\left(\rho\mathbf{V}^+\cdot\rho\mathbf{V}^+\right).
+    """
+
+    def __init__(self, boundary_pressure=101325):
+        """Initialize the boundary condition object."""
+        self._pressure = boundary_pressure
+        PrescribedFluidBoundary.__init__(
+            self, boundary_state_func=self.outflow_state
+        )
+
+    def outflow_state(self, discr, btag, gas_model, state_minus, **kwargs):
+        """Get the exterior solution on the boundary.
+
+        This is the partially non-reflective boundary state described by
+        [Mengaldo_2014]_ eqn. 40 if super-sonic, 41 if sub-sonic.
+        """
+        actx = state_minus.array_context
+        nhat = thaw(discr.normal(btag), actx)
+        # boundary-normal velocity
+        boundary_vel = np.dot(state_minus.velocity, nhat)*nhat
+        boundary_speed = actx.np.sqrt(np.dot(boundary_vel, boundary_vel))
+        speed_of_sound = state_minus.speed_of_sound
+        kinetic_energy = gas_model.eos.kinetic_energy(state_minus.cv)
+        gamma = gas_model.eos.gamma(state_minus.cv, state_minus.temperature)
+        external_pressure = 2*self._pressure - state_minus.pressure
+        boundary_pressure = actx.np.where(actx.np.greater(boundary_speed,
+                                                          speed_of_sound),
+                                          state_minus.pressure, external_pressure)
+        internal_energy = boundary_pressure / (gamma - 1)
```

Comment:
This line only works for IdealGas. For mixture, it is necessary to use "gas_model.eos.internal_energy"

Reply:
I have not made this change yet because the boundary routines seem inconsistent with the reference we are using.   The ideal vs. mixture part is OK; it's the part where we seem to prescribe a "pressure" for the outflow that's bugging me.

Reply:
Fixed with https://github.com/illinois-ceesd/mirgecom/pull/705